### ISSUE TITLE
Server segfaults when a client connects

### ISSUE THREAD

**Issue by oznogon** [daid/EmptyEpsilon#769]
*Regression*

On a build from the master branches of both EmptyEpsilon and SeriousProton, the server segfaults when a client connects.

Reproduced on the master branches in Windows 10 and macOS 10.15.

This does not affect builds from the EE-2020.02.18 tag of both repos.

*Steps to reproduce*

1. Start an instance of EE to be the server
2. Click "Start Server"
3. Leave default scenario values and click "Start Scenario"
4. Start another instance of EE to be the client
5. In the client instance, click "Start Client"
6. The server is autodiscovered on the LAN. Click it in the server list
7. Click "Connect"

*Expected behavior*

The client connects to the server.

*Observed behavior*

The server immediately crashes.

End of the client's console/log output prior to the crash:

```
[INFO]  ServerScanner::New server: 192.168.0.21 35666 Server
[INFO]  Loaded: gui/ButtonBackground.active with 0 sprites
[INFO]  GameClient: Connected, waiting for authentication
Failed to bind socket to port 35667
```

macOS report excerpt:

```
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000060
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [55094]

VM Regions Near 0x60:
--> 
    __TEXT                 0000000107764000-0000000107e7d000 [ 7268K] r-x/r-x SM=COW  /Users/USER/*/EmptyEpsilon.app/Contents/MacOS/EmptyEpsilon

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0                                 	0x0000000107ccb464 TcpSocket::send(sf::Packet&) + 20
1                                 	0x0000000107ced276 GameServer::update(float) + 3510
2                                 	0x0000000107d83abb Engine::runMainLoop() + 1243
3                                 	0x000000010776ba02 main + 15922
4   libdyld.dylib                 	0x00007fff71f707fd start + 1
```

Windows report excerpt:

```
EmptyEpsilon.exe caused an Access Violation at location 005BDB46 in module EmptyEpsilon.exe Reading from location 00000038.

Registers:
eax=00000000 ebx=10295e98 ecx=00000000 edx=00000000 esi=10208c40 edi=016f3598
eip=005bdb46 esp=0148eb60 ebp=0148eb88 iopl=0         nv up ei pl nz na po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00010206

AddrPC   Params
005BDB46 0148EC8C 0078AF23 0178D4F0  EmptyEpsilon.exe!TcpSocket::send
005C71E7 00000000 007AD570 007ACCEC  EmptyEpsilon.exe!GameServer::update
00619511 0148FB98 0148FBC8 0076833A  EmptyEpsilon.exe!Engine::runMainLoop
00404155 015DD4D0 00000010 00000003  EmptyEpsilon.exe!main
004013E3 00356000 76EE6340 0148FFDC  EmptyEpsilon.exe!__tmainCRTStartup  [./mingw-w64-crt/crt/crtexe.c @ 334]
76EE6359 00356000 4D735AF3 00000000  KERNEL32.DLL!@BaseThreadInitThunk@12
77087B74 FFFFFFFF 770A8F3C 00000000  ntdll.dll!__RtlUserThreadStart
77087B44 004014E0 00356000 00000000  ntdll.dll!__RtlUserThreadStart@8
```

**Comment by oznogon** [comment]
Full crash logs:

[EmptyEpsilon.RPT.txt](https://github.com/daid/EmptyEpsilon/files/4300967/EmptyEpsilon.RPT.txt)

**Comment by daid** [comment]
Got a fix for this, but it on a computer I cannot access till monday (forgot to push it)

**Comment by daid** [comment]
I think the change in https://github.com/daid/SeriousProton/commit/ca54a06e5e48e207f47641788d9a702cb1351d45
Will fix this. Specifically the change in `src/multiplayer_server.cpp` on line 166/172

**Comment by oznogon** [comment]
That does seem to fix it, thanks @daid!

**Comment by amir-arad** [comment]
> I think the change in [daid/SeriousProton@ca54a06](https://github.com/daid/SeriousProton/commit/ca54a06e5e48e207f47641788d9a702cb1351d45)
> Will fix this. Specifically the change in `src/multiplayer_server.cpp` on line 166/172

could you please explain why? It seems lilke the proxy is not yet in use anywhere and most of the changes only concern the proxy commands. is it the moving of `clientList.push_back`? if so, why does that fix the problem? is it the `std::move` ?

**Comment by daid** [comment]
The bug was introduced in this patch:
https://github.com/daid/SeriousProton/commit/548be980232eda676e2a949adc78636ebadbd0d9

Which moved the TcpSocket into a unique_ptr. Which keeps track of the memory management, at the "cost" that only a single reference is allowed to exist to it. So the std::move is needed to move that reference. Now, the mistake I made was still using it after it was moved, this corrupted things as "info.socket" was used after the move, but the move made it invalid. Operating on invalid data is always bad, as it becomes unpredictable, in this case, crashes happen on disconnect of the client.
